### PR TITLE
PADV-1187: Refactor resource link launch code and improve code quality

### PR DESCRIPTION
**Pull request by kuipumu**
## Ticket

https://agile-jira.pearson.com/browse/PADV-1187

## Description

This PR add a refactor to the resource link launch code, by moving the existing code for resource link launches into a new `resource_link_launch` module, additionally this PR improves the code quality (more specifically on docstrings and comments) of the refactored code, a minor improvement on the verification of the launch message 'message_type' value on the `ResourceLinkLaunchView` and a rename of the `ResourceLinkLaunchView` `get_resource_launch` method to `get_launch_response`.

## Type of Change

- [x] Move existing resource link launch code to `resource_link_launch` module.
- [x] Improve the verification of the launch message 'message_type' value on the `ResourceLinkLaunchView` post method.
- [x] Improve code quality of the `resource_link_launch` module code.

## Testing:

- Run the LMS: `make dev.up.lms`.
- Run the learning MFE on the host machine: `npm run start`.
- Install `openedx_lti_tool_plugin` on the LMS.
- Run ngrok or any other similar service on LMS: `ngrok http 18000`
- Run ngrok or any other similar service on learning MFE: `ngrok http 2000`
- Add required settings to LMS.
	
```
# Enable LTI Tool Provider Plugin.
OLTITP_ENABLE_LTI_TOOL = True

# Cookies settings.
SHARED_COOKIE_DOMAIN = ".ngrok.io"  # Replace .ngrok.io with the domain of the service used to expose the LMS and learning MFE if using another service other than ngrok.
# MFE config API settings.
ENABLE_MFE_CONFIG_API = True
MFE_CONFIG_API_CACHE_TIMEOUT = 0
```

- Create a site for your external LMS address: http://localhost:18000/admin/sites/site/add/.
- Create a site configuration for your site: http://localhost:18000/admin/site_configuration/siteconfiguration/add/

```
  {
      "SESSION_COOKIE_DOMAIN": ".ngrok.io", # This should be the same value has SHARED_COOKIE_DOMAIN setting.
      "LEARNING_MICROFRONTEND_URL": "https://learning-mfe.ngrok.io",
      "MFE_CONFIG": {
          "LMS_BASE_URL": "https://lms.ngrok.io",
          "LOGIN_URL": "https://lms.ngrok.io/login",
          "LOGOUT_URL": "https://lms.ngrok.io/logout",
          "MARKETING_SITE_BASE_URL": "https://lms.ngrok.io",
          "BASE_URL": "lms.ngrok.io",
          "LEARNING_BASE_URL": "https://learning-mfe.ngrok.io,
          "REFRESH_ACCESS_TOKEN_ENDPOINT": "https://lms.ngrok.io/login_refresh"
      }
  }
```

- Restart the LMS: `make dev.restart-container.lms`.
- Modify the learning MFE `webpack.dev.config.js` file (create one if not file exists.):

```
const path = require('path');
const { createConfig } = require('@edx/frontend-build');
const config = createConfig('webpack-dev', {
    devServer: {
      allowedHosts: 'all',
    },
});
config.resolve.modules = [
  path.resolve(__dirname, './src'),
  'node_modules',
];
module.exports = config;
```

- Modify the learning MFE `.env.development` file and add these variables:

```
APP_ID='learning'
MFE_CONFIG_API_URL='https://lms.ngrok.io/api/mfe_config/v1'
```

- Create a new LTI 1.3 tool key config: http://localhost:18000/admin/lti1p3_tool_config/ltitoolkey/add/ (You can use IMS RI to create a key pair: https://lti-ri.imsglobal.org/keygen/index)
- Create a new LTI 1.3 tool config: http://localhost:18000/admin/lti1p3_tool_config/ltitool/add/

```
Issuer: https://saltire.lti.app/platform
Client id: saltire.lti.app
Auth login: url: https://saltire.lti.app/platform/auth
Auth token url: https://saltire.lti.app/platform/token/sda42bb0cf2352259a367a404d48d54e8
Key set url: https://saltire.lti.app/platform/jwks/sda42bb0cf2352259a367a404d48d54e8
Deployment ids: ["cLWwj9cbmkSrCNsckEFBmA"]
```

- Go to the saLTIre platform https://saltire.lti.app/platform
- Go to "Security Model" on the left sidebar and set these settings:

```
Message URL: https://lms.ngrok.io/openedx_lti_tool_plugin/1.3/launch/{course_id}
Initiate login URL: https://lms.ngrok.io/openedx_lti_tool_plugin/1.3/login
Redirection URI(s): https://lms.ngrok.io/openedx_lti_tool_plugin/1.3/launch/{course_id}
Public keyset URL: https://lms.ngrok.io/openedx_lti_tool_plugin/1.3/pub/jwks
```

- Click on the "Save" button on the top navbar.
- Click on the dropdown next to the "Connect" button on the top navbar.
- Click on "Open in iframe".
- The launch should redirect you to the requested course on the learning MFE.

## Reviewers

- [x] @alexjmpb  
- [x] @anfbermudezme

### CODE REVIEW

File: openedx_lti_tool_plugin/resource_link_launch/ags/tasks.py [hunk 1]
```diff
@@ -8,12 +13,13 @@
 
 from openedx_lti_tool_plugin.edxapp_wrapper.grades_module import course_grade_factory
 from openedx_lti_tool_plugin.edxapp_wrapper.modulestore_module import modulestore
-from openedx_lti_tool_plugin.models import LtiGradedResource
+from openedx_lti_tool_plugin.resource_link_launch.ags.models import LtiGradedResource
 
 log = logging.getLogger(__name__)
+MODULE_PATH = 'openedx_lti_tool_plugin.tasks'
```
Comment (alexjmpb) [inline openedx_lti_tool_plugin/resource_link_launch/ags/tasks.py:19]: What do you think of defining this variable in the __init__.py

**Review by alexjmpb — DISMISSED** [review]

File: openedx_lti_tool_plugin/resource_link_launch/views.py [hunk 2]
```diff
@@ -0,0 +1,409 @@
...
+        aud: Union[list, str],
+        sub: str,
+        pii: dict,
+    ) -> LtiProfile:
+        """Get LtiProfile instance.
+
+        Get LtiProfile instance and update instance PII if instance is not created.
+
+        Args:
+            iss: Issuer claim.
```
Comment (anfbermudezme) [inline openedx_lti_tool_plugin/resource_link_launch/views.py:210]: @kuipumu Same comment of references.

**Review by anfbermudezme — CHANGES_REQUESTED** [review]

File: openedx_lti_tool_plugin/resource_link_launch/ags/__init__.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+"""LTI Assignment and Grade Services.
```
[the remark this answers is not in the document]
Comment (kuipumu) [reply openedx_lti_tool_plugin/resource_link_launch/ags/__init__.py:1]: @anfbermudezme I could add references to the specification on this module docstring and also update the README to also include those references there, what else do you think we could add to make easier for developers to have a better understanding of this module and AGS?

File: openedx_lti_tool_plugin/resource_link_launch/views.py [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+
+        return redirect('render_xblock', usage_key_string)
+
+    def handle_ags(  # pylint: disable=inconsistent-return-statements
+        self,
+        launch_message: DjangoMessageLaunch,
+        launch_data: dict,
+        lti_profile: LtiProfile,
+        context_key: str,
+    ):
```
[the remark this answers is not in the document]
Comment (kuipumu) [reply openedx_lti_tool_plugin/resource_link_launch/views.py:371]: @anfbermudezme I will keep the return type hint the same, the method doesn't return the `LtiGradedResource` instance, and since the method always returns None the type hint is not necessary, however I noticed the return on the docstring was incorrect so I will fix that

File: openedx_lti_tool_plugin/resource_link_launch/exceptions.py [hunk 5]
```diff
@@ -1,4 +1,4 @@
-"""Exceptions for openedx_lti_tool_plugin."""
+"""Exceptions."""
```
[the remark this answers is not in the document]
Comment (kuipumu) [reply openedx_lti_tool_plugin/resource_link_launch/exceptions.py:1]: @anfbermudezme I think is redundant and also these exceptions aren't for the whole `openedx_lti_tool_plugin` module but rather the `resource_link_launch` module, also if we had to change the `openedx_lti_tool_plugin` module in the future it will be easier to have unnecessary references to such name like in this docstring.

File: openedx_lti_tool_plugin/resource_link_launch/ags/tasks.py [hunk 6]
```diff
@@ -8,12 +13,14 @@
 
 from openedx_lti_tool_plugin.edxapp_wrapper.grades_module import course_grade_factory
 from openedx_lti_tool_plugin.edxapp_wrapper.modulestore_module import modulestore
-from openedx_lti_tool_plugin.models import LtiGradedResource
+from openedx_lti_tool_plugin.resource_link_launch.ags import MODULE_PATH
+from openedx_lti_tool_plugin.resource_link_launch.ags.models import LtiGradedResource
 
 log = logging.getLogger(__name__)
+MODULE_PATH = f'{MODULE_PATH}.tasks'
```
[the remark this answers is not in the document]
Comment (kuipumu) [reply openedx_lti_tool_plugin/resource_link_launch/ags/tasks.py:20]: @anfbermudezme What specific comment where you referring to?

File: openedx_lti_tool_plugin/resource_link_launch/ags/validators.py [hunk 7]
```diff
@@ -0,0 +1,34 @@
+"""Validators."""
+from django.core.exceptions import ValidationError
+from django.utils.translation import gettext as _
+from opaque_keys import InvalidKeyError
+from opaque_keys.edx.keys import CourseKey, UsageKey
+
+
+def validate_context_key(value: str):
```
[the remark this answers is not in the document]
Comment (kuipumu) [reply openedx_lti_tool_plugin/resource_link_launch/ags/validators.py:8]: @anfbermudezme This method doesn't set a variable it's a Django validator so it's not necessary to return a value

File: openedx_lti_tool_plugin/resource_link_launch/views.py [hunk 8]
```diff
@@ -0,0 +1,409 @@
...
+from openedx_lti_tool_plugin.resource_link_launch.ags.models import LtiGradedResource
+from openedx_lti_tool_plugin.resource_link_launch.exceptions import LtiToolLaunchException
+from openedx_lti_tool_plugin.utils import get_client_id, get_pii_from_claims
+from openedx_lti_tool_plugin.views import LtiToolBaseView
+from openedx_lti_tool_plugin.waffle import ALLOW_COMPLETE_COURSE_LAUNCH, COURSE_ACCESS_CONFIGURATION, SAVE_PII_DATA
+
+log = logging.getLogger(__name__)
+
+AGS_CLAIM_ENDPOINT = 'https://purl.imsglobal.org/spec/lti-ags/claim/endpoint'
+AGS_SCORE_SCOPE = 'https://purl.imsglobal.org/spec/lti-ags/scope/score'
```
[the remark this answers is not in the document]
Comment (kuipumu) [reply openedx_lti_tool_plugin/resource_link_launch/views.py:40]: @anfbermudezme I responded on this previous comment: https://github.com/Pearson-Advance/openedx-lti-tool-plugin/pull/33#discussion_r1592627272

File: openedx_lti_tool_plugin/resource_link_launch/views.py [hunk 9]
```diff
@@ -0,0 +1,409 @@
...
+        Raises:
+            LtiToolLaunchException: If CourseAccessConfiguration instance
+                does not exist for LtiTool or the `course_id` is not allowed.
+
+        """
+        if not COURSE_ACCESS_CONFIGURATION.is_enabled():
+            return
+
+        lti_tool_config = self.tool_config.get_lti_tool(iss, aud)
+        course_access_config = CourseAccessConfiguration.objects.filter(
```
[the remark this answers is not in the document]
Comment (kuipumu) [reply openedx_lti_tool_plugin/resource_link_launch/views.py:186]: @anfbermudezme I think it could be better but I prefer not to address this on this PR since it will require adding more changes that aren't a priority for this PR

File: openedx_lti_tool_plugin/resource_link_launch/ags/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (anfbermudezme) [reply openedx_lti_tool_plugin/resource_link_launch/ags/__init__.py:1]: @kuipumu Probably in the Readme in a subsection adds a little description of these concepts. Thanks !

File: openedx_lti_tool_plugin/resource_link_launch/ags/signals.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+from opaque_keys.edx.keys import CourseKey
+
+from openedx_lti_tool_plugin.edxapp_wrapper.core_signals_module import course_grade_changed
+from openedx_lti_tool_plugin.edxapp_wrapper.grades_module import problem_weighted_score_changed
+from openedx_lti_tool_plugin.models import LtiProfile, UserT
+from openedx_lti_tool_plugin.resource_link_launch.ags.models import LtiGradedResource
+from openedx_lti_tool_plugin.resource_link_launch.ags.tasks import send_problem_score_update, send_vertical_score_update
+from openedx_lti_tool_plugin.utils import is_plugin_enabled
+
+MAX_SCORE = 1.0
```
[the remark this answers is not in the document]
Comment (anfbermudezme) [reply openedx_lti_tool_plugin/resource_link_launch/ags/signals.py:25]: Righ, t but you have more constants throught ags directory. It's just a suggention of organization. Something like this https://github.com/Pearson-Advance/course_operations/blob/master/pearson_course_operation/constants.py. But I agree with your decision.

File: openedx_lti_tool_plugin/resource_link_launch/ags/tests/test_tasks.py [hunk 11]
```diff
@@ -222,18 +197,13 @@ def test_send_problem_score_update(
...
         log_mock: MagicMock,
         lti_graded_resource_mock: MagicMock,
     ):
-        """Test send_problem_score_update task without graded resources.
-
-        Args:
-            log_mock: mocked log.
-            lti_graded_resource_mock: mocked LtiGradedResource model.
-        """
+        """Test without graded resources."""
```
Comment (anfbermudezme) [inline openedx_lti_tool_plugin/resource_link_launch/ags/tests/test_tasks.py:206]: @kuipumu Why did you remove the description of the arguments?

File: openedx_lti_tool_plugin/resource_link_launch/ags/tests/__init__.py [hunk 12]
```diff
@@ -0,0 +1,3 @@
+"""Test ags module."""
+
+MODULE_PATH = 'openedx_lti_tool_plugin.resource_link_launch.ags'
```
Comment (anfbermudezme) [inline openedx_lti_tool_plugin/resource_link_launch/ags/tests/__init__.py:3]: @kuipumu You are setting this MODULE_PATH in each testing module. What do you think move it into the constants module or in an more general __init__.py as a dict to obtain the the desire path calling it dfor example `MODULE_PATH.get('ags_path')` (It is just an example.)

File: openedx_lti_tool_plugin/resource_link_launch/tests/__init__.py [hunk 13]
```diff
@@ -0,0 +1,3 @@
+"""Test resource_link_launch module."""
+
+MODULE_PATH = 'openedx_lti_tool_plugin.resource_link_launch'
```
Comment (anfbermudezme) [inline openedx_lti_tool_plugin/resource_link_launch/tests/__init__.py:3]: @kuipumu Same comment as before.

File: openedx_lti_tool_plugin/tests/test_urls.py [hunk 14]
```diff
@@ -2,34 +2,22 @@
...
 
 
-class TestUrls(TestCase):
-    """Test openedx_lti_tool_plugin URL configuration."""
+class TestUrlPatterns(TestCase):
+    """Test Django URL configuration."""
 
-    def test_lti_tool_login_url_resolves(self):
-        """Test LtiToolLoginView URL can be resolved."""
+    def test_oidc_login_view_urls(self):
```
Comment (anfbermudezme) [inline openedx_lti_tool_plugin/tests/test_urls.py:11]: @kuipumu What is the meaning of oidc?

File: openedx_lti_tool_plugin/utils.py [hunk 15]
```diff
@@ -46,7 +48,8 @@ def get_client_id(aud: list, azp: Optional[str]) -> str:
 def get_pii_from_claims(claims: dict) -> dict:
     """Get PII from claims dictionary.
 
-    This function extracts PII from a dictionary of claims.
+    This function extracts PII (Personal Identifiable Information)
```
Comment (anfbermudezme) [inline openedx_lti_tool_plugin/utils.py:51]: @kuipumu I like this docstring. :)

**Review by anfbermudezme — CHANGES_REQUESTED** [review]

File: openedx_lti_tool_plugin/resource_link_launch/ags/tests/test_tasks.py (hunk 11 above)
Comment (kuipumu) [reply openedx_lti_tool_plugin/resource_link_launch/ags/tests/test_tasks.py:206]: @anfbermudezme I removed them because it's very clear that they are all mock objects, also I think this affects the readability of the tests

File: openedx_lti_tool_plugin/tests/test_urls.py (hunk 14 above)
Comment (kuipumu) [reply openedx_lti_tool_plugin/tests/test_urls.py:11]: @anfbermudezme This is an abbreviation of "OpenID Connect", this is a reference to the authorization endpoint of the LTI 1.3 launch: https://openid.net/specs/openid-connect-core-1_0.html#AuthorizationEndpoint

I left this change by mistake, I was going to rename the LtiToolLoginView to OIDCLoginView but instead kept the name but this test remained with the new name of the view, I will change it back to `test_lti_tool_login_view_urls`

File: openedx_lti_tool_plugin/resource_link_launch/ags/tests/__init__.py (hunk 12 above)
Comment (kuipumu) [reply openedx_lti_tool_plugin/resource_link_launch/ags/tests/__init__.py:3]: @anfbermudezme I think for now I will keep it by setting a MODULE_PATH on each module that requires it because I'm not sure right now what could be the best approach to improve this and also this PR main focus was on relocating the resource link launch code, improving the tests or code quality was an extra and could be improved in another iteration, I will think of another approach to this based on your suggestion so we can apply it to new and existing tests on a future ticket

**Review by anfbermudezme — APPROVED** [review]
